src/Logic.hs

```diff
@@ -669,14 +699,35 @@ handleBuildStatusChanged buildSha newStatus state = pure $
 -- same commit hash depending on the CI you are using with GitHub.
 -- This guarantees that 'BuildSucceeded' or 'BuildFailed' statuses
 -- cannot be overridden by later builds of a branch pointing to the same hash.
-supersedes :: BuildStatus -> BuildStatus -> Bool
-newStatus `supersedes` oldStatus        | sameStatus newStatus oldStatus
-                                        = False -- url change, ignore
-_         `supersedes` (BuildFailed _)  = False -- final status
-_         `supersedes` BuildSucceeded   = False -- final status
-_         `supersedes` _                = True
-
--- | Compares if two build statuses are the same
+supersedes :: Sha -> Maybe Check -> BuildStatus -> BuildStatus -> PullRequest -> Bool
+supersedes sha maybeCheck newStatus oldStatus pr =
+  -- There's probably a neater way to do this. But the semantics are quite
+  -- complicated.
+  -- - If a known check was submitted with the pull request, check if we already
+  --   received a previous status for it, and avoid overwriting that status.
+  -- - If we didn't receive a known check, verify if the check is in the
+  --   mandatory list of checks we know of.
+  -- - If no known check was submitted, check the previous integration status of
+  --   the sha and avoid overwriting that status if necessary.
+  let maybeCheckStatus = do
+        perShaChecks <- Map.lookup sha (Pr.integrationChecks pr)  -- final status
+        check <- maybeCheck
+        Map.lookup check perShaChecks
+      ignoreStatus oldCheckStatus | sameStatus newStatus oldCheckStatus = False
+      ignoreStatus oldCheckStatus = isFinalStatus oldCheckStatus
+      isFinalStatus oldCheckStatus = case oldCheckStatus of
+        BuildFailed _ -> False
+        BuildSucceeded -> False
+        _ -> True
+  in case maybeCheckStatus of
+    Just checkStatus -> ignoreStatus checkStatus
+    Nothing -> case maybeCheck of
+      -- We don't have a status stored for this (mandatory) check
+      -- Add it, the status is final
+      Just _ -> not (isFinalStatus newStatus)
```

Comment:
If the status correspond to a check, and the new status is final, it won't supersede the non-existing previous status?

I'm probably missing something here